src/nbl/asset/utils/CHLSLCompiler.cpp

```diff
@@ -8,14 +8,230 @@
+    // TODO support this
+    //res_str = IShaderCompiler::encloseWithinExtraInclGuards(std::move(res_str), _maxInclCnt, name.string().c_str());
+
+    return new tcpp::StringInputStream(std::move(res_str));
+}
+
+class DxcCompilationResult
+{
+public:
+    ComPtr<IDxcBlobEncoding> errorMessages;
+    ComPtr<IDxcBlob> objectBlob;
+    ComPtr<IDxcResult> compileResult;
+
+    char* GetErrorMessagesString()
+    {
+        return reinterpret_cast<char*>(errorMessages->GetBufferPointer());
+    }
+};
+
+DxcCompilationResult dxcCompile(const CHLSLCompiler* compiler, std::string& source, LPCWSTR* args, uint32_t argCount, const CHLSLCompiler::SOptions& options)
+{
+    if (options.genDebugInfo)
+    {
+        std::ostringstream insertion;
+        insertion << "// commandline compiler options : ";
+
+        std::wstring_convert<std::codecvt_utf8<wchar_t>, wchar_t> conv;
+        for (uint32_t arg = 0; arg < argCount; arg ++)
+        {
+            auto str = conv.to_bytes(args[arg]);
+            insertion << str.c_str() << " ";
+        }
+
+        insertion << "\n";
+        compiler->insertIntoStart(source, std::move(insertion));
+    }
+    
+    ComPtr<IDxcBlobEncoding> src;
+    auto res = compiler->getDxcUtils()->CreateBlob(reinterpret_cast<const void*>(source.data()), source.size(), CP_UTF8, &src);
+    assert(SUCCEEDED(res));
+
+    DxcBuffer sourceBuffer;
+    sourceBuffer.Ptr = src->GetBufferPointer();
+    sourceBuffer.Size = src->GetBufferSize();
+    sourceBuffer.Encoding = 0;
+
+    ComPtr<IDxcResult> compileResult;
+    res = compiler->getDxcCompiler()->Compile(&sourceBuffer, args, argCount, nullptr, IID_PPV_ARGS(compileResult.GetAddressOf()));
+    // If the compilation failed, this should still be a successful result
+    assert(SUCCEEDED(res));
+
+    HRESULT compilationStatus = 0;
+    res = compileResult->GetStatus(&compilationStatus);
+    assert(SUCCEEDED(res));
+
+    ComPtr<IDxcBlobEncoding> errorBuffer;
+    res = compileResult->GetErrorBuffer(errorBuffer.GetAddressOf());
+    assert(SUCCEEDED(res));
+
+    DxcCompilationResult result;
+    result.errorMessages = errorBuffer;
+    result.compileResult = compileResult;
+    result.objectBlob = nullptr;
+
+    if (!SUCCEEDED(compilationStatus))
+    {
+        options.preprocessorOptions.logger.log(result.GetErrorMessagesString(), system::ILogger::ELL_ERROR);
+        return result;
+    }
+
+    ComPtr<IDxcBlob> resultingBlob;
+    res = compileResult->GetResult(resultingBlob.GetAddressOf());
+    assert(SUCCEEDED(res));
+
+    result.objectBlob = resultingBlob;
+
+    return result;
+}
+
+
+std::string CHLSLCompiler::preprocessShader(std::string&& code, IShader::E_SHADER_STAGE& stage, const SPreprocessorOptions& preprocessOptions) const
+{
+    if (preprocessOptions.extraDefines.size())
+    {
+        insertExtraDefines(code, preprocessOptions.extraDefines);
+    }
+    if (preprocessOptions.includeFinder != nullptr)
```

Comment:
this if statement contains much more than it shoud, if there is no includeFinder then don't resolve/replace the includes